### PR TITLE
Add users by QR code

### LINKED ISSUE #2025: QR-codes for faster contact sharing
# The problem
When I want to share a contact (my or someone else's) with the person I've met face to face, I need to spell (or show) my full Matrix ID, which is not very convenient. Otherwise, I can send `matrix.to` link with some external communication method, which is also not very fast.

# Proposed solution
On the "profile page" add "show QR-code" button, which will create QR-code with `matrix.to` link.

It will allow users to quickly exchange Matrix IDs, which seems quite usefull.

### PR DESCRIPTION
**Pull request by TR-SLimey**
Second attempt (#2276) including all fixes from review of last PR.

Fixes #2025

Also fixes a small typo in RoomDirectoryItem.kt

### Pull Request Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [x] Pull request is based on the develop branch
- [x] Pull request updates [CHANGES.md](https://github.com/vector-im/element-android/blob/develop/CHANGES.md)
- [x] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)

<details>
<summary>
Screenshots
</summary>
Sharing user by QR code (Light theme)

image

Sharing user by QR code (Dark theme)

image

</details>

Tested and working:
- Request permissions before opening scanner
- Cleanly close scanner if permissions not granted
- Open scanner if permissions granted (even after denied previously)
- Open DM if it already exists
- Create DM if it does not
- Let user know that they cannot DM themselves if they try
- Show progress while creating DM
- Tell the user if QR code is not valid
- Generate valid QR codes in user profile

Signed-off-by: TR_SLimey | tr_slimey@protonmail.com | @tr_slimey:an-atom-in.space

### CODE REVIEW

File: vector/src/main/res/drawable/ic_fab_add_by_mxid.xml
Comment (TR-SLimey) [inline vector/src/main/res/drawable/ic_fab_add_by_mxid.xml:1]: This file will need to be replaced as it's just a placeholder for an actual icon. Perhaps a magnifying glass for search?

File: vector/src/main/res/drawable/ic_fab_add_by_qr_code.xml
Comment (TR-SLimey) [inline vector/src/main/res/drawable/ic_fab_add_by_qr_code.xml:1]: This file will also need to be replaced. Perhaps a QR code with a plus sign in the corner?

File: vector/src/main/res/layout/motion_dms_fab_menu_merge.xml
Comment (TR-SLimey) [inline vector/src/main/res/layout/motion_dms_fab_menu_merge.xml:76]: This may need to be replaced, depending on whether a plus sign on this FAB is desired. It looks makes the add group button look a little odd though.

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (TR-SLimey) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:44]: I have no idea what this block here does, since I just copy+pasted it, but it made the code work so I left it in. This can probably be cleaned up but I'll need some tips as to how.

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:60]: Shouldn't we call `scannerView.setResultHandler(null)` for symmetry?

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:105]: You should move this code (all the `else` block) in the viewModel, by adding an action in `CreateDirectRoomAction`

File: vector/src/main/res/drawable/ic_fab_add_by_mxid.xml
Comment (bmarty) [inline vector/src/main/res/drawable/ic_fab_add_by_mxid.xml:15]: This is a really long path for a small icon...

File: vector/src/main/res/drawable/ic_fab_add_by_qr_code.xml
Comment (bmarty) [inline vector/src/main/res/drawable/ic_fab_add_by_qr_code.xml:6]: This is identical with `ic_fab_add_by_mxid.xml`. I do not know where the icon comes from. Maybe I can ask to our design team to provide an alternative icon for qr code

**Review by bmarty — CHANGES_REQUESTED** [review]
Hello, thanks for the update.
The PR is really great! I have some few remark, and especially one about moving the treatment into the ViewModel. Feel free to ask if you need help, or if you prefer us to do the change.
QR code dialog and QR code scanning tested OK on my device.
Thanks again!

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (TR-SLimey) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:105]: I think the whole if/else can be removed if/when #2342 gets merged, right? In that case creating a new DM will automatically switch to the old one anyway, so there will be no need to check that here. Should I still move the else block or wait for #2342 to get merged, rebase, and then remove the whole thing?

File: vector/src/main/res/drawable/ic_fab_add_by_mxid.xml
Comment (TR-SLimey) [reply vector/src/main/res/drawable/ic_fab_add_by_mxid.xml:15]: Both this and the below are just copies of `ic_fab_add_members.xml`. I've used that as a placeholder to avoid having to change the filename in code when changing for actual icons, but yeah, I think some new icons should be designed and these files replaced.

File: vector/src/main/res/drawable/ic_fab_add_by_qr_code.xml
Comment (TR-SLimey) [reply vector/src/main/res/drawable/ic_fab_add_by_qr_code.xml:6]: See [above](https://github.com/vector-im/element-android/pull/2289#discussion_r518279698)

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (TR-SLimey) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:60]: I have no idea :P but I changed it

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (TR-SLimey) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:38]: I'm actually having a bit of trouble with that. I tried removing both this and the `@CallSuper` block, which gives me undefined errors for session. I then put `private val session: Session` in `constructor()` like some other fragments have, but then I get runtime crashes with the following stack trace:

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{im.vector.app.debug/im.vector.app.features.createdirect.CreateDirectRoomActivity}: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment im.vector.app.features.createdirect.CreateDirectRoomByQrCodeFragment: could not find Fragment constructor
```

Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt]: To fix the pb with the Fragment, you should declare it in the file FragmentModule.kt. But this is not necessary if you do the following:
The `session` should stay private in the ViewModel, and you should use `VectorViewModelAction` and `VectorViewEvents` to communicate between the ViewModel and the Fragment.

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:95]: #2342 will be merged soon, I will see after what we can do here.

File: vector/src/main/res/drawable/ic_fab_add_by_mxid.xml
Comment (bmarty) [reply vector/src/main/res/drawable/ic_fab_add_by_mxid.xml:15]: ok, fine, thanks

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:42]: This is not what you want to do :). Please revert.
What I meant is: `scannerView.setResultHandler(null)` has to be added in the `onPause()` method.

Comment (TR-SLimey) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:42]: Yeah, I've replaced the wrong thing, just noticed that like 10 minutes ago and replaced it, but didn't push yet :P

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (TR-SLimey) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt:91]: I've just updated my branch and removed the whole if statement. It seems to work as desired.

File: vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt
Comment (TR-SLimey) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt]: I've had a look at it but still can't figure it out. Maybe it would be better if I left it as is and you guys could change it after merging?

Comment (TR-SLimey) [reply vector/src/main/java/im/vector/app/features/createdirect/CreateDirectRoomByQrCodeFragment.kt]: Thanks. Then I'll mark this as resolved

**Comment by TR-SLimey** [comment]
Hey, @bmarty I noticed this PR is still sitting around, anything I can do to help it along or is it blocked by something?

**Comment by BillCarsonFr** [comment]
Hello @TR-SLimey , I took your commitz and put them here https://github.com/vector-im/element-android/pull/2417
In a more generic PR to improve invite.
Closing this PR in favor of that one

**Pull request closed without merging**
